### ISSUE TITLE
Add actions in relationGroup

### ISSUE THREAD

**Issue by sterling9** [udos86/ng-dynamic-forms#335]
Hello,

For dynamic-form-control-relation.model, shall we add more actions such as 'show' and 'hide' to enrich the logic other than "Disable" and "Enable"?

Thanks

**Comment by udos86** [comment]
@sterling9 Thanks for the proposal. 

As soon as Kendo and Material are stable I'll try to do a refactoring of the relation feature.

**Comment by Palad1NPT** [comment]
that would be awesome, any idea of how could i implement this feature? i really need it for a project

**Comment by Palad1NPT** [comment]
I currently have hacked the @ng2-dynamic-forms folder in my project, to do this i edited the:
./node_modules/@ng2-dynamic-forms/core/bundles/core.es.js

I've made the following changes:

- After var DYNAMIC_FORM_CONTROL_ACTION_ENABLE = "ENABLE" added the following code:
	```
var DYNAMIC_FORM_CONTROL_ACTION_HIDE = "HIDE";
	var DYNAMIC_FORM_CONTROL_ACTION_SHOW = "SHOW";
```

- After function DynamicFormRelationService.prototype.findActivationRelation = function (relGroups) created this function:
	```
DynamicFormRelationService.prototype.findActivationRelationHidden = function (relGroups) {
        return relGroups.find(function (rel) { return rel.action === DYNAMIC_FORM_CONTROL_ACTION_HIDE || rel.action === DYNAMIC_FORM_CONTROL_ACTION_SHOW ; });
    };
```

- After function DynamicFormRelationService.prototype.isFormControlToBeDisabled = function (relGroup, formGroup) created this function:
```
DynamicFormRelationService.prototype.isFormControlToBeHidden = function (relGroup, formGroup) {
        return relGroup.when.reduce(function (toBeHidden, rel, index) {
            var control = formGroup.get(rel.id);
            if (control && relGroup.action === DYNAMIC_FORM_CONTROL_ACTION_HIDE) {
                if (index > 0 && relGroup.connective === DYNAMIC_FORM_CONTROL_CONNECTIVE_AND && !toBeHidden) {
                    return false;
                }
                if (index > 0 && relGroup.connective === DYNAMIC_FORM_CONTROL_CONNECTIVE_OR && toBeHidden) {
                    return true;
                }
                return rel.value === control.value || rel.status === control.status;
            }
            if (control && relGroup.action === DYNAMIC_FORM_CONTROL_ACTION_SHOW) {
                if (index > 0 && relGroup.connective === DYNAMIC_FORM_CONTROL_CONNECTIVE_AND && toBeHidden) {
                    return true;
                }
                if (index > 0 && relGroup.connective === DYNAMIC_FORM_CONTROL_CONNECTIVE_OR && !toBeHidden) {
                    return false;
                }
                return !(rel.value === control.value || rel.status === control.status);
            }
            return false;
        }, false);
    };
```
	
- After function  DynamicFormControlComponent.prototype.updateModelDisabled = function (relation) created the following:
    ```
DynamicFormControlComponent.prototype.updateModelHidden = function (relation) {
        this.model.hiddenUpdates.next(this.relationService.isFormControlToBeHidden(relation, this.group));
    };
```
	
- After function DynamicFormControlComponent.prototype.onModelDisabledUpdates = function (value) created the following:
	```
DynamicFormControlComponent.prototype.onModelHiddenUpdates = function (value) {
        value ? this.model.cls.grid.container = "hidden" : this.model.cls.grid.container = "ui-grid-row";
    };
```

- ON DynamicFormControlComponent.prototype.setControlRelations added the following code:
	```
relActivation = null;
	relActivation = this.relationService.findActivationRelationHidden(this.model.relation);
	if (relActivation) {
		this.updateModelHidden(relActivation);
		this.relationService.getRelatedFormControls(this.model, this.group).forEach(function (control) {
			_this.subscriptions.push(control.valueChanges.subscribe(function () { return _this.updateModelHidden(relActivation); }));
			_this.subscriptions.push(control.statusChanges.subscribe(function () { return _this.updateModelHidden(relActivation); }));
		});
	}
```
		

- ON  DynamicFormControlComponent.prototype.ngOnInit added the following code right after this.subscriptions.push(this.model.disabledUpdates.subscribe(this.onModelDisabledUpdates.bind(this))):
	`this.subscriptions.push(this.model.hiddenUpdates.subscribe(this.onModelHiddenUpdates.bind(this)));   `
	

- ON var DynamicFormControlModel = (function () added the following code at the end:
	```
this.hiddenUpdates = new Subject();
	this.hiddenUpdates.subscribe(function (value) { return value ? _this.cls.grid.container = "hidden" : _this.cls.grid.container = "ui-grid-row"; });
```

- After Fuction DynamicFormControlComponent.prototype.onModelDisabledUpdates = function (value) created:
	```
DynamicFormControlComponent.prototype.onModelHiddenUpdates = function (value) {
        value ? this.model.cls.grid.container = "hidden" : this.model.cls.grid.container = "ui-grid-row";
    };
```

- Last line export added:
`DYNAMIC_FORM_CONTROL_ACTION_ENABLE, DYNAMIC_FORM_CONTROL_ACTION_HIDE,`

Limitations:
- Currently i'm hard coding control.cls.grid.container class to "ui-grid-row" when visible.
- As soon as i update my dependecies i have to do this again.
-If i need another action is a lot of code to replicate and will have the same limitations.

Hope @udos86 can work on a permanent and more elegant solution, due to project timeline i can't put more effort on this.

**Comment by stevehoper** [comment]
Thanks for the proposal, I will try to refactor it on basic controls and hopefully @udos86 can release a prototype version soon.

**Comment by ticxx** [comment]
Hi
More wishes for the relation feature.
It would also be very nice if the control could be disabled and emptied. 
For example, in our forms we have combined checkboxes with textareas where users can leave special comments. If the checkbox is set to false, the comment should be deleted.

Thanks for all your work!

**Comment by udos86** [comment]
@ticxx Hi!

> Thanks for all your work!

You're welcome!

Right now I'm collecting all proposals for this library feature to make up a plan fo refactoring.

**Comment by Orrefors** [comment]
Hi @Palad1NPT  

I've been looking at your example but i have the "this.model.cls" as undefined.
I've seen this example with cls on older thread, is this gone now?

So this doesnt work:
```
    DynamicFormControlComponent.prototype.onModelHiddenUpdates = /**
    * @param {?} value
    * @return {?}
    */
    function (value) {
        value ? this.model.cls.grid.container = "hidden" : this.model.cls.grid.container = "ui-grid-row";
    };
```


```
id: "drawingLandDone",
            initialCount: 1,
            relation: [
                {
                    action: "HIDE",
                    when: [
                        {
                            id: "isDrawingLandDone",
                            value: true
                        }
                    ]
                }
            ]

```

**Comment by cab-peax** [comment]
@udos86 Hi Udo, any news on this feature? "Show" and "Hide" would be really great to have.

**Comment by seanforyou23** [comment]
+1 this feature would be very helpful. I see `VISIBLE` and `HIDDEN` are options in dynamic-form-control-relation.model.d.ts, but they do not seem to work in the same basic sense that enable/disable do. Is there a recommended way to work around this in the short term while the larger refactoring is being planned?

**Comment by janekdijkstra** [comment]
Hey, any updates on this?

**Comment by stevehoper** [comment]
@udos86 
Hi, any updates on this please?

**Comment by udos86** [comment]
_major refactoring will be finally available in `8.0.0`_
